src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+        /// Set grove output pin low or high
+        /// </summary>
+        /// <param name="port">The Groove Pin, can be any combination of Groove1Pin1, Groove1Pin2, Groove2Pin1 and/or Groove2Pin2</param>
+        /// <param name="state">The pin state. false for low or true for high.</param>
+        public void SetGrooveState(GroovePort port, bool state)
+        {
+            byte[] outArray = { SpiAddress, (byte)SpiMessageType.SetGrooveState, (byte)port, (byte)(state ? 1 : 0) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Set grove output pin PWM
+        /// </summary>
+        /// <param name="port">The Groove Pin, can be any combination of Groove1Pin1, Groove1Pin2, Groove2Pin1 and/or Groove2Pin2</param>
+        /// <param name="duty">The PWM duty cycle in percent from 0.0 to 100.0, 1 floating point precision</param>
+        public void SetGroovePwmDuty(GroovePort port, double duty)
+        {
+            duty = Math.Clamp(duty, (byte)0, (byte)100);
+
+            var duty_value = (UInt16)(duty * 10.0);
+            byte[] outArray = { SpiAddress, (byte)SpiMessageType.SetGroovePwmDuty, (byte)port, (byte)((duty_value >> 8) & 0xFF), (byte)(duty_value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Set grove PWM frequency
+        /// </summary>
+        /// <param name="port">The grove port(s).GROVE_1 and / or GROVE_2.</param>
+        /// <param name="freq">The PWM frequency.Range is 3 through 48000Hz.Default is 24000(24kHz). Limit to 48000, which is the highest frequency supported for 0.1% resolution.</param>
+        public void GetGroovePwmFrequency(GroovePort port, uint freq = 24000)
+        {
+            freq = Math.Clamp(freq, 3, 48000);
+            byte[] outArray = { SpiAddress, (byte)SpiMessageType.SetGroovePwmFrequency, (byte)port, (byte)((freq >> 8) & 0xFF), (byte)(freq & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Conduct an I2C transaction
+        /// </summary>
+        /// <param name="port">The grove port. GROVE_1 or GROVE_2</param>
+        /// <param name="addr">The I2C address of the slave to be addressed.</param>
+        /// <param name="arrayToSend">An array of bytes to send.</param>
+        /// <param name="inBytes">The number of bytes to read.</param>
+        /// <returns>Returns a byte array with what has been read from the I2C element</returns>
+        public byte[] GrooveI2cTransfer(GroovePort port, byte addr, byte[] arrayToSend, byte inBytes = 0)
+        {
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            var timeout = stopwatch.ElapsedMilliseconds + 5;
+            while (true)
+            {
+                try
+                {
+                    GrooveI2cStart(port, addr, arrayToSend, inBytes);
+                    break;
+                }
+                catch (Exception ex) when (ex is IOException)
```

Comment:
since you only check for one exception type this can be just `catch (IOException ex)`